tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/EnergyTests.java

```diff
@@ -0,0 +1,118 @@
+            assertEquals(3, simpleHandler.insert(3, transaction));
+            assertEquals(8, simpleHandler.getAmountAsLong());
+
+            transaction.commit();
+        }
+
+        assertEquals(8, simpleHandler.getAmountAsLong());
+    }
+
+    @Test
+    void testItemAccessEnergyHandler() {
+        // Starting items: 2 diamonds. Diamond will be considered the energy container here.
+        var startingItems = NonNullList.of(ItemStack.EMPTY, new ItemStack(Items.DIAMOND, 2));
+        var handler = new ItemStacksResourceHandler(startingItems) {
+            @Override
+            protected int getCapacity(int index, ItemResource resource) {
+                // Limit to 2
+                return Math.min(2, super.getCapacity(index, resource));
+            }
+        };
+        var itemAccess = ItemAccess.forHandlerIndex(handler, 0);
+
+        // Create the energy handler
+        var energyHandler = new ItemAccessEnergyHandler(itemAccess, TestMod.ENERGY.get(), 60, 50, 30);
+
+        try (Transaction transaction = Transaction.open(null)) {
+            // Insertion of 200 should only insert 100 (50 per item).
+            assertEquals(100, energyHandler.insert(200, transaction));
+            assertEquals(50, itemAccess.getResource().get(TestMod.ENERGY));
+            // Insertion of 200 should only insert 20 (10 per item) due to the capacity.
+            assertEquals(20, energyHandler.insert(200, transaction));
+            assertEquals(60, itemAccess.getResource().get(TestMod.ENERGY));
+            // Extraction of 30 should extract 30 (15 per item).
+            assertEquals(30, energyHandler.extract(30, transaction));
+            assertEquals(45, itemAccess.getResource().get(TestMod.ENERGY));
+            // Check amount and capacity.
+            assertEquals(90, energyHandler.getAmountAsLong());
+            assertEquals(120, energyHandler.getCapacityAsLong());
+
+            // Now check that everything returns 0 if we change the item in the slot.
+            handler.set(0, ItemResource.EMPTY, 0);
+            ensureEmpty(energyHandler, transaction);
+        }
+    }
+
+    private static void ensureEmpty(EnergyHandler handler, TransactionContext transaction) {
+        assertEquals(0, handler.insert(Integer.MAX_VALUE, transaction));
+        assertEquals(0, handler.extract(Integer.MAX_VALUE, transaction));
+        assertEquals(0, handler.getAmountAsLong());
+        assertEquals(0, handler.getCapacityAsLong());
+    }
+
+    @Mod(TestMod.MOD_ID)
+    public static class TestMod {
+        private static final String MOD_ID = "energy_test";
+        private static final DeferredRegister.DataComponents components = DeferredRegister.createDataComponents(Registries.DATA_COMPONENT_TYPE, MOD_ID);
+        private static final Supplier<DataComponentType<Integer>> ENERGY = components.registerComponentType("energy",
+                b -> b.persistent(Codec.INT).networkSynchronized(ByteBufCodecs.INT));
```

Comment:
I know it is a test but given people might reference it for their own impls, please use `ExtraCodecs.NON_NEGATIVE_INT` (ideally we would limit it to the actual amount instead of just disallowing negatives, but I don't think that is worth it for a test). And use `VAR_INT` for the sync codec as energy can never be negative so there isn't any concern that it will make it bigger on average.